class-tgm-plugin-activation.php

```diff
@@ -1128,6 +1108,75 @@ public function _get_plugin_data_from_name( $name, $data = 'slug' ) {
 		}
 
 		/**
+		 * Retrieve the download url for a package.
+		 *
+		 * @since 2.5.0
+		 *
+		 * @param string $slug
+		 *
+		 * @return string Plugin download URL or path to local file
+		 */
+		public function get_download_url( $slug ) {
+			$dl = '';
+
+			if ( ! empty( $this->plugins[ $slug ]['source'] ) ) {
+
+				$source = $this->plugins[ $slug ]['source'];
+
+				// Is this a WP repo plugin ?
+				if ( 'repo' === $source || preg_match( self::WP_REPO_REGEX, $source ) ) {
+
+					$dl = $this->get_wp_repo_download_url( $slug );
+				}
+				// Is this an external package url ?
+				elseif ( preg_match( self::EXT_REPO_REGEX, $source ) ) {
+
+					$dl = $source;
+				}
+				// This must be a bundled/pre-packaged plugin. Prefix it with the default path.
+				else {
+					$dl = $this->default_path . $source;
+				}
+			}
+
+			return $dl;
+		}
+
+		/**
+		 * Retrieve the download url for a WP repo package.
+		 *
+		 * @since 2.5.0
+		 *
+		 * @param string $slug Plugin slug
+		 *
+		 * @return string Plugin download URL
+		 */
+		protected function get_wp_repo_download_url( $slug ) {
+			
+			if ( ! function_exists( 'plugins_api' ) ) {
+				require_once ABSPATH . 'wp-admin/includes/plugin-install.php'; // Need for plugins_api.
+			}
+
+			$source = '';
+
+			// Try to grab information from WordPress API
+			$api = plugins_api( 'plugin_information', array( 'slug' => $slug, 'fields' => array( 'sections' => false ) ) );
+			if ( is_wp_error( $api ) ) {
+				if ( true === WP_DEBUG ) {
```

Comment:
My code should be no different - just refactoring out multiple calls to wp_die(), esc_html() and array value access.